src/responsibleai/rai_analyse/generate_pdf.py

```diff
@@ -0,0 +1,1226 @@
+                div(
+                    metric_section,
+                    _class="nobreak_div"
+                )
+            )
+
+        left_container = div(left_elems, _class="left")
+
+        def get_fairness_bar_plot(data):
+            y_data = [str(c) + "<br>" + str(int(100 * data[c]["population"])) + "%" for c in data]
+            x_data = [100*(get_metric("selection_rate", data[c]["y_test"], data[c]["y_pred"])) for c in data]
+            x_data = [[x, 100-x] for x in x_data]
+
+            tickvals = [0, 25, 50, 75, 100]
+            ticktext = [str(x) + "%" for x in tickvals]
+
+            png_base64 = get_bar_plot(y_data, x_data,
+                                      tickvals=tickvals, ticktext=ticktext, tickappend="%")
+
+            return div(img(_src="data:image/png;base64,{}".format(png_base64)), _class="image_div")
+
+        def get_table_row(heading, data):
+            table_row_elems = []
+            table_row_elems.append(th(heading, _class="header_cell"))
+            for v in data:
+                table_row_elems.append(td(v, _class="cell"))
+            return tr(table_row_elems, _class="row")
+
+        def get_table(data):
+            metric_list = [d for d in data["metrics"]]
+            #metric_list = [m for sublist in metric_list for m in sublist]
+
+            horizontal_headings = [d.replace("_", "<br>") for d in metric_list]
+            vertical_headings = list(data["statistics"].keys())
+
+            headings_td = [td(_class="header_cell")] + [td(x, _class="header_cell") for x in horizontal_headings]
+            headings = thead(tr(headings_td, _class="row"), _class="table-head")
+
+            rows_elems = []
+            for vh in vertical_headings:
+                row_data = []
+                for m in metric_list:
+                    row_data.append(
+                        round(data["metrics"][m]["group_metric"][vh], 2)
+                    )
+                rows_elems.append(get_table_row(
+                    vh,
+                    row_data)
+                )
+
+            body = tbody(rows_elems, _class="table-body")
+
+            return table(headings, body, _class="table")
+
+        main_elems = []
+        # Selection rate
+        for f in data:
+            main_elems.append(
+                div(
+                    h2("Feature \"{}\"".format(f)),
+                    h3("Selection rate"),
+                    get_fairness_bar_plot(data[f]["statistics"]),
+                    _class="nobreak_div"
+                )
+            )
+
+            main_elems.append(
+                div(
+                    h3("Analysis across cohorts"),
+                    get_table(data[f]),
+                    _class="nobreak_div"
+                )
+            )
+
+        main_container = div(main_elems, _class="main")
+
+        return str(div(get_page_divider("Fairness"), left_container, main_container, _class="container nobreak_div"))
+
+
+class RegressionComponents:
```

Comment:
Can the individual classes be put into their own files. At 1200 lines, my eyes are glazing over.

Reply:
Particularly since.... have I missed instance methods on this class?

Reply:
Resolved with refactoring.